[motion-path] Have basic shapes use an offset containing block rect

nmoucht opened the pull request:
#### f6fd650e3830ebe58caebefa3914dd243b19f0ae
<pre>
[motion-path] Have basic shapes use an offset containing block rect
<a href="https://bugs.webkit.org/show_bug.cgi?id=261217">https://bugs.webkit.org/show_bug.cgi?id=261217</a>
rdar://115068196

Reviewed by Tim Nguyen.

Update basic shapes to use an offset containing block rect in
MotionPath::computePathForShape.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/MotionPath.cpp:
(WebCore::offsetRectForData):
(WebCore::MotionPath::computePathForBox):
(WebCore::MotionPath::computePathForShape):
* Source/WebCore/rendering/MotionPath.h:
* Source/WebCore/rendering/PathOperation.h:

Canonical link: <a href="https://commits.webkit.org/267866@main">https://commits.webkit.org/267866@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7c427da1cefbd41d10adeea850dc27c7c3b69281

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17830 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18156 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18702 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19657 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16676 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18025 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21448 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18302 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18691 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18045 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18314 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15484 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20524 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15542 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16238 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22784 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16558 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16406 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20649 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16979 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14367 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16081 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/16079 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4267 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20439 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16827 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/MotionPath.cpp:
```diff
@@ -194,19 +194,36 @@ std::optional<Path> MotionPath::computePathForRay(const RayPathOperation& rayPat
+        auto& shape = pathOperation.basicShape();
+        auto rect = motionPathData->containingBlockBoundingRect.rect();
+        if (is<BasicShapeInset>(shape))
+            rect = offsetRectForData(*motionPathData).rect();
```
I don't think it's just inset(), it's all the shapes that should be unaffected by properties adding more offsets from the container. Sadly we don't have testcases exercising this bug on other shapes

nmoucht replied:
Right. Switched to all shapes except for path.

nt1m commented on Source/WebCore/rendering/MotionPath.cpp:
```diff
@@ -194,19 +194,36 @@ std::optional<Path> MotionPath::computePathForRay(const RayPathOperation& rayPat
+        auto& shape = pathOperation.basicShape();
+        auto rect = motionPathData->containingBlockBoundingRect.rect();
+        if (!is<BasicShapePath>(shape))
+            rect = offsetRectForData(*motionPathData).rect();
```
Excluding path seems arbitrary to me. I would expect the behavior to be consistent for all shapes. Can we check what Firefox & Chrome do here?

nmoucht replied:
Yeah interestingly offsetting the rect causes the path tests to fail. Gonna try and understand why.

nt1m reviewed (commented):
Can we change the commit title? Adding some tests for non-inset shapes would be good too.

nt1m reviewed (approved):
Looks fine, but please undo any unnecessary test changes!

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-shape-ellipse-006.html:
```diff
@@ -0,0 +1,34 @@
+<!-- You can increase the numbers to handle anti-aliasing -->
+<meta name="fuzzy" content="maxDifference=0-150; totalPixels=0-530">
+<link rel="author" title="Daniil Sakhapov" href="sakhapov@google.com">
+<link rel="match" href="offset-path-shape-ellipse-003-ref.html">
```
If the ref is different than `offset-path-shape-ellipse-003-ref.html` please reflect it here. Otherwise if it's identical to `offset-path-shape-ellipse-003-ref.html`, please delete the `006-ref.html` file you added :)

nmoucht replied:
Fixed.

nmoucht committed:
[motion-path] Have basic shapes use an offset containing block rect
https://bugs.webkit.org/show_bug.cgi?id=261217
rdar://115068196

Reviewed by Tim Nguyen.

Update basic shapes to use an offset containing block rect in
MotionPath::computePathForShape.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/MotionPath.cpp:
(WebCore::offsetRectForData):
(WebCore::MotionPath::computePathForBox):
(WebCore::MotionPath::computePathForShape):
* Source/WebCore/rendering/MotionPath.h:
* Source/WebCore/rendering/PathOperation.h:

Canonical link: https://commits.webkit.org/267866@main
